Archive/Imo/Imo1982Q3.lean

```diff
@@ -0,0 +1,381 @@
+      linarith
+      rw [lt_sub_iff_add_lt, zero_add, Nat.one_lt_cast]
+      apply hN
+      rw [@nonempty_range_iff]
+      exact ne_zero
+    have sum_fin_pos : 0 < ∑ n : Fin N, x (↑n + 1) := by
+      apply Finset.sum_pos; intro i _hi
+      apply x_pos (i +1)
+      rw [@univ_nonempty_iff, ← @Fin.pos_iff_nonempty]
+      linarith
+    convert_to
+      (N - 1) / N * (1 / ∑ n in range (N - 1), x (n + 1)) ≤ 1 / (∑ n : Fin N, x (n + 1)) using 3
+    rw [sum_range]
+    convert_to 1 / (N * (∑ n in range (N - 1), x (n + 1)) / (N - 1)) ≤ 1 / (∑ n : Fin N, x (n + 1))
+    field_simp
+    convert_to 1 / ∑ i ∈ range (N - 1), x (i + 1) * ↑N / (↑N - 1) ≤ 1 / (∑ n : Fin N, x (n + 1))
+    rw [mul_comm, sum_mul, sum_div]
+    rw [div_le_div_iff (mul_sum_pos) (sum_fin_pos), one_mul, one_mul, ]
+    calc ∑ n : Fin N, x (↑n + 1) = ∑ n in range N, x (n + 1) := by rw [sum_range]
+    _ = ∑ n in range (N - 1 + 1), x (n + 1) := by
+      rw [Nat.sub_one_add_one_eq_of_pos (by linarith [hN])]
+    _ = ∑ n in range (N - 1), x (n + 1) + x N := by
+      rw [sum_range_succ, Nat.sub_one_add_one_eq_of_pos (by linarith [hN])]
+    _ ≤ ∑ n in range (N - 1), x (n + 1) + (∑ n ∈ range (N - 1), x (n + 1)) / (↑N - 1) := by
+      apply add_le_add_left; rw [@sum_range]; apply ineq₁ hx N hN;
+    _ = ∑ n in range (N - 1), x (n + 1) + (∑ n ∈ range (N - 1), x (n + 1) / (↑N - 1)) :=  by
+      rw [sum_div]
+    _ = ∑ n in range (N - 1), (x (n + 1) + x (n + 1) / (↑N - 1)) := by rw [Finset.sum_add_distrib]
+    _ = ∑ n in range (N - 1),  N * x (n + 1) / (↑N - 1) := by
+      apply Finset.sum_congr (by rfl)
+      intro n _hn
+      nth_rewrite 1 [
+        ← one_mul (x (n + 1)),
+        ← div_self (a := (N - 1 : ℝ)) (ne_zero'),
+        mul_comm,
+        mul_div,
+        div_add_div_same
+        ]
+      nth_rewrite 2 [← mul_one (x (n + 1))]
+      rw [← mul_add, mul_comm]
+      simp only [sub_add_cancel]
+    _ = ∑ i ∈ range (N - 1), x (i + 1) * ↑N / (↑N - 1) := by
+      apply Finset.sum_congr (by rfl); intro n _hn; rw [mul_comm]
+
+
+lemma ineq₃ {x : ℕ → ℝ} (x_pos : ∀ i, x i > (0 : ℝ)) :
+    ∀ N > 1, 2 * (∑ n : Fin N, x (n + 1)) ≤ 1 + (∑ n : Fin N, x (n + 1))^2 := by
+    intro N hN
+    have sum_fin_pos : 0 < ∑ n : Fin N, x (↑n + 1) := by
+      apply Finset.sum_pos
+      intro i _hi
+      apply x_pos (i +1)
+      rw [@univ_nonempty_iff, ← @Fin.pos_iff_nonempty]
+      linarith
+    calc
+    2 * (∑ n : Fin N, x (n + 1)) = 2 * (1^(1/2 : ℝ) * ((∑ n : Fin N, x (n  + 1))^2)^(1/2 : ℝ)) := by
+      rw [one_rpow, one_mul, ← Real.sqrt_eq_rpow, sqrt_sq _]
+      apply le_of_lt sum_fin_pos
+    _ ≤ 2 * ((1/2 : ℝ) * 1 + (1/2 : ℝ) * (∑ n : Fin N, x (n  + 1))^2) := by
+      rw [mul_le_mul_left (by norm_num)]
+      apply Real.geom_mean_le_arith_mean2_weighted
+        (by norm_num) (by norm_num) (by norm_num) (sq_nonneg _) (by norm_num)
+    _ ≤ 1 + (∑ n : Fin N, x (n  + 1))^2 := by field_simp
+
+lemma Ico_sdiff_zero_eq_Ico : ∀ N, Ico 0 N \ {0} = Ico 1 N := by
+    intro N
```

Comment:
General comments applicable both to this lemma and elsewhere in this PR:

* Normally rather than having a lemma statement start with `∀` after the colon and then the proof starting with `intro`, we use named hypotheses before the colon, so `(N : ℕ)` in this case, and so don't need to start the proof with `intro`. Much the same applies if the statement is of the form `h → p`: move the hypotheses `h` before the colon rather than having an implication on the right hand side of the colon.
* Standard indentation of the proof body is two columns, not four.

Specific remark about this lemma: it's essentially a combination of the existing lemmas `sdiff_singleton_eq_erase`, `Ico_erase_left` and `Ico_succ_left`. That should allow a much shorter proof. It might then be reasonable to inline that at the point where this lemma is needed.